File: packages/python/pomlight/inline.py
from __future__ import annotations

import re
from .xml_parser import XmlNode, XmlElement
from .types import State
from .tags import normalize_tag
from .expr import interpolate, interpolate_pre, eval_condition, escapes


def render_inline(children: list[XmlNode], state: State) -> str:
    out = ""
    for c in children:
        if isinstance(c, str):
            t = interpolate(c, state.ctx)
            t = escapes(t)
            t = re.sub(r"\s+", " ", t)  # collapse whitespace
            out += t
        else:
            # if-directive on inline elements
            if_attr = c.attrs.get("if")
            if if_attr is not None:
                if not eval_condition(if_attr, state.ctx):
                    continue
            inner = render_inline(c.children, state).strip()
            tag = normalize_tag(c.tag)
            if tag == "b":
                out += f"**{inner}**"
            elif tag == "i":
                out += f"*{inner}*"
            elif tag == "u":
                out += f"__{inner}__"
            elif tag in ("s", "strike"):
                out += f"~~{inner}~~"
            elif tag == "code":
                out += f"`{inner}`"
            elif tag == "br":
                nl_count = int(c.attrs.get("newLineCount", "1"))
                out += "\n" * nl_count
            elif tag == "audio":
                pass
            elif tag == "img":
                img_syn = c.attrs.get("syntax")
                img_mm = img_syn == "multimedia" or (not img_syn and "alt" not in c.attrs)
                if not img_mm and "alt" in c.attrs:
                    out += interpolate(c.attrs["alt"], state.ctx)
            else:  # span or unknown
                ws = c.attrs.get("whiteSpace") or c.attrs.get("white-space")
                if ws == "trim":
                    pre = render_inline_pre(c.children, state).strip()
                    out += pre
                elif ws == "pre":
                    out += render_inline_pre(c.children, state)
                else:
                    out += inner
    return out


def render_inline_pre(children: list[XmlNode], state: State) -> str:
    out = ""
    for c in children:
        if isinstance(c, str):
            t = interpolate_pre(c, state.ctx)
            t = escapes(t)
            out += t
        else:
            if_attr = c.attrs.get("if")
            if if_attr is not None:
                if not eval_condition(if_attr, state.ctx):
                    continue
            inner = render_inline_pre(c.children, state)
            ws = c.attrs.get("whiteSpace") or c.attrs.get("white-space")
            tag = normalize_tag(c.tag)
            if tag == "b":
                out += f"**{inner}**"
            elif tag == "i":
                out += f"*{inner}*"
            elif tag == "code":
                out += f"`{inner}`"
            elif tag == "br":
                out += "\n"
            elif tag == "span":
                if ws == "trim":
                    out += inner.strip()
                else:
                    out += inner
            else:
                out += inner
    return out

File: packages/deno/src/inline.ts
import type { XmlNode } from "./xml.ts";
import type { State } from "./types.ts";
import { normalizeTag } from "./tags.ts";
import { interpolate, interpolatePre, evalCondition, escapes } from "./expr.ts";

// ---------------------------------------------------------------------------
// Inline rendering  (text, <b>, <code>, <span>)
// ---------------------------------------------------------------------------

export function renderInline(children: XmlNode[], state: State): string {
  let out = "";
  for (const c of children) {
    if (typeof c === "string") {
      let t = interpolate(c, state.ctx);
      t = escapes(t);
      t = t.replace(/\s+/g, " "); // collapse whitespace
      out += t;
    } else {
      // if-directive on inline elements
      if (c.attrs.if !== undefined) {
        if (!evalCondition(c.attrs.if, state.ctx)) continue;
      }
      const inner = renderInline(c.children, state).trim();
      switch (normalizeTag(c.tag)) {
        case "b":
          out += `**${inner}**`;
          break;
        case "i":
          out += `*${inner}*`;
          break;
        case "u":
          out += `__${inner}__`;
          break;
        case "s":
        case "strike":
          out += `~~${inner}~~`;
          break;
        case "code":
          out += `\`${inner}\``;
          break;
        case "br": {
          const nlCount = parseInt(c.attrs.newLineCount ?? "1", 10);
          out += "\n".repeat(nlCount);
          break;
        }
        case "audio":
          break;
        case "img": {
          const imgSyn = c.attrs.syntax;
          const imgMm = imgSyn === "multimedia" || (!imgSyn && !c.attrs.alt);
          if (!imgMm && c.attrs.alt) out += interpolate(c.attrs.alt, state.ctx);
          break;
        }
        default: { // span or unknown → pass through
          const ws = c.attrs.whiteSpace ?? c.attrs["white-space"];
          if (ws === "trim") {
            const pre = renderInlinePre(c.children, state).trim();
            out += pre;
          } else if (ws === "pre") {
            out += renderInlinePre(c.children, state);
          } else {
            out += inner;
          }
          break;
        }
      }
    }
  }
  return out;
}

/** Render inline children preserving whitespace (pre mode) — no collapsing. */
export function renderInlinePre(children: XmlNode[], state: State): string {
  let out = "";
  for (const c of children) {
    if (typeof c === "string") {
      let t = interpolatePre(c, state.ctx);
      t = escapes(t);
      // No whitespace collapsing in pre mode
      out += t;
    } else {
      if (c.attrs.if !== undefined) {
        if (!evalCondition(c.attrs.if, state.ctx)) continue;
      }
      const inner = renderInlinePre(c.children, state);
      const ws = c.attrs.whiteSpace ?? c.attrs["white-space"];
      switch (normalizeTag(c.tag)) {
        case "b": out += `**${inner}**`; break;
        case "i": out += `*${inner}*`; break;
        case "code": out += `\`${inner}\``; break;
        case "br": out += "\n"; break;
        case "span":
          if (ws === "trim") out += inner.trim();
          else out += inner;
          break;
        default: out += inner; break;
      }
    }
  }
  return out;
}
